datadog_lambda/tracing.py

```diff
@@ -248,28 +249,55 @@ def extract_context_from_sqs_or_sns_event_or_context(event, lambda_context):
-            dd_json_data = dd_payload.get(
-                "stringValue",
-                dd_payload.get("Value", r"{}"),
-            )
+        if dd_payload:
+            # SQS uses dataType and binaryValue/stringValue
+            # SNS uses Type and Value
+            dd_json_data = None
+            dd_json_data_type = dd_payload.get("Type", dd_payload.get("dataType", ""))
+            if dd_json_data_type == "Binary":
+                dd_json_data = dd_payload.get(
+                    "binaryValue",
+                    dd_payload.get("Value", r"{}"),
+                )
+                dd_json_data = base64.b64decode(dd_json_data)
+            elif dd_json_data_type == "String":
+                dd_json_data = dd_payload.get(
+                    "stringValue",
+                    dd_payload.get("Value", r"{}"),
+                )
+            else:
+                logger.debug(
+                    "Datadog Lambda Python only supports extracting trace"
+                    "context from String or Binary SQS/SNS message attributes"
+                )
+
+            if dd_json_data:
+                dd_data = json.loads(dd_json_data)
+                return propagator.extract(dd_data)
         else:
-            logger.debug(
-                "Datadog Lambda Python only supports extracting trace"
-                "context from String or Binary SQS/SNS message attributes"
-            )
-            return extract_context_from_lambda_context(lambda_context)
-        dd_data = json.loads(dd_json_data)
-        return propagator.extract(dd_data)
+            # Handle case where trace context is injected into attributes.AWSTraceHeader
+            # example: Root=1-654321ab-000000001234567890abcdef;Parent=0123456789abcdef;Sampled=1
+            x_ray_header = first_record.get("attributes", {}).get("AWSTraceHeader")
+            if x_ray_header:
+                x_ray_context = parse_xray_header(x_ray_header)
+                trace_id_parts = x_ray_context.get("trace_id", "").split("-")
+                if len(trace_id_parts) > 2 and trace_id_parts[2].startswith("00000000"):
+                    # If it starts with eight 0's padding,
+                    # then this AWSTraceHeader contains Datadog injected trace context
+                    logger.debug(
+                        "Found dd-trace injected trace context from AWSTraceHeader"
+                    )
+                    dd_data = {
+                        TraceHeader.TRACE_ID: str(
+                            int(trace_id_parts[2][8:], 16)
+                        ),  # remove padding and convert the hex str to a decimal str
+                        TraceHeader.PARENT_ID: str(
+                            int(x_ray_context["parent_id"], 16)
+                        ),  # convert the hex str to a decimal str
+                        TraceHeader.SAMPLING_PRIORITY: x_ray_context["sampled"],
+                    }
+                    return propagator.extract(dd_data)
```

Comment:
Instead of calling extract, why not just manually create the `Context` object? For example, this code here will not work if the extract style is just `w3c`.

Reply:
Updated it and tested a testing layer with the java upstream lambda having `DD_TRACE_PROPAGATION_STYLE_INJECT = tracecontext` and the downstream python lambda having `DD_TRACE_PROPAGATION_STYLE_EXTRACT=tracecontext`. 
<img width="1329" alt="Screenshot 2024-03-27 at 2 52 58 PM" src="https://github.com/DataDog/datadog-lambda-python/assets/5253430/12d004a5-bec7-448c-ad26-f43e9f7f62ff">